producer/cloudwatch_logs.go

```diff
@@ -0,0 +1,164 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package producer
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/cloudwatchlogs"
+	"github.com/trivago/gollum/core"
+	"time"
+)
+
+/*
+AWS CloudWatch specific constants.
+Also see http://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/cloudwatch_limits_cwl.html
+*/
+const (
+	// Maximum number of log events in a batch.
+	maxBatchEvents = 10000
+	// Maximum batch size in bytes.
+	maxBatchSize = 1048576
+	// Maximum event size in bytes.
+	maxEventSize = 262144
+	// A batch of log events in a single PutLogEvents request cannot span more than 24 hours.
+	maxBatchTimeSpan = 24 * time.Hour
+	// How many bytes to append to each log event.
+	eventSizeOverhead = 26
+	// DescribeLogStreams transactions/second.
+	describeLogstreamsDelay = 200 * time.Millisecond
+	// PutLogEvents 5 requests/second/log stream.
+	putLogEventsDelay = 200 * time.Millisecond
+)
+
+type CloudwatchLogs struct {
+	core.BufferedProducer `gollumdoc:"embed_type"`
+	stream                *string
+	group                 *string
+	token                 *string
+	service               *cloudwatchlogs.CloudWatchLogs
+}
+
+func init() {
+	core.TypeRegistry.Register(CloudwatchLogs{})
+}
+
+// Configure initializes this producer with values from a plugin config.
+func (prod *CloudwatchLogs) Configure(conf core.PluginConfigReader) error {
+	prod.SetStopCallback(prod.flush)
+	stream := conf.GetString("stream", "")
+	if stream == "" {
+		return fmt.Errorf("stream name can not be empty")
+	}
+	group := conf.GetString("group", "")
+	if group == "" {
+		return fmt.Errorf("group name can not be empty")
+	}
+	prod.stream = &stream
+	prog.group = &group
+	return nil
+}
+
+// Put log events and update sequence token.
+// Possible errors http://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_PutLogEvents.html
+func (prod *CloudwatchLogs) upload() {
+	logevents := make([]*cloudwatchlogs.InputLogEvent, 0, len(events))
+	for _, elem := range events {
+		logevents = append(logevents, &cloudwatchlogs.InputLogEvent{
+			Message:   aws.String(elem.msg),
+			Timestamp: aws.Int64(elem.timestamp),
+		})
+	}
+	params := &cloudwatchlogs.PutLogEventsInput{
+		LogEvents:     logevents,
+		LogGroupName:  prod.group,
+		LogStreamName: prod.stream,
+		SequenceToken: prod.token,
+	}
+	// When rejectedLogEventsInfo is not empty, app can not
+	// do anything reasonable with rejected logs. Ignore it.
+	// Meybe expose some statistics for rejected counters.
+	resp, err := dst.svc.PutLogEvents(prod.cloudwatchLogsParams)
+	if err == nil {
+		prod.cloudwatchLogsParams.SequenceToken = resp.NextSequenceToken
+	}
+}
+
+func (prod *Console) Produce(workers *sync.WaitGroup) {
+	defer prod.WorkerDone()
+	prod.AddMainWorker(workers)
+	prod.MessageControlLoop(prod.printMessage)
```

Comment:
printMessage function does not exist.
You probably want to call `upload` here?
If you use the buffered producer you will only get single messages though.
If you want to send batches of messages you should have a look at `core.BatchProducer`